lib/mayaUsd/fileio/primUpdaterManager.cpp

```diff
@@ -1724,4 +1747,270 @@ bool PrimUpdaterManager::readPullInformation(const MDagPath& dagPath, Ufe::Path&
     return false;
 }
 
+/* static */
+std::list<PrimUpdaterManager::VariantSetDescriptor>
+PrimUpdaterManager::variantSetDescriptors(const Ufe::Path& p)
+{
+    std::list<VariantSetDescriptor> vsd;
+    auto                            path = p;
+    while (path.nbSegments() > 1) {
+        auto ancestor = Ufe::Hierarchy::createItem(path);
+        auto usdAncestor = std::static_pointer_cast<MayaUsd::ufe::UsdSceneItem>(ancestor);
+        auto variantSets = usdAncestor->prim().GetVariantSets();
+        std::list<VariantSelection> vs;
+        for (const auto& vsn : variantSets.GetNames()) {
+            vs.emplace_back(vsn, variantSets.GetVariantSelection(vsn));
+        }
+        vsd.emplace_back(path, vs);
+        path = path.pop();
+    }
+    return vsd;
+}
+
+void PrimUpdaterManager::beginManagePulledPrims()
+{
+    TF_VERIFY(_pulledPrims.root()->empty());
+    _orphanedNodesManager = std::make_shared<OrphanedNodesManager>(_pulledPrims);
+    Ufe::Scene::instance().addObserver(_orphanedNodesManager);
+
+    // Observe Maya so we can stop scene observation on file new or open.
+    MStatus                status;
+    MSceneMessage::Message msgs[] = { MSceneMessage::kBeforeNew, MSceneMessage::kBeforeOpen };
+    for (auto msg : msgs) {
+        _fileCbs.append(MSceneMessage::addCallback(msg, beforeNewOrOpenCallback, this, &status));
+        CHECK_MSTATUS(status);
+    }
+}
+
+void PrimUpdaterManager::endManagePulledPrims()
+{
+    TF_VERIFY(_orphanedNodesManager);
+    TF_VERIFY(Ufe::Scene::instance().removeObserver(_orphanedNodesManager));
+    auto status = MMessage::removeCallbacks(_fileCbs);
+    CHECK_MSTATUS(status);
+    _fileCbs.clear();
+    _orphanedNodesManager.reset();
+    _pulledPrims.clear();
+}
+
+/*static*/
+void PrimUpdaterManager::beforeNewOrOpenCallback(void* clientData)
+{
+    auto* pum = static_cast<PrimUpdaterManager*>(clientData);
+    pum->endManagePulledPrims();
+}
+
+//==============================================================================
+//
+// CLASS PrimUpdaterManager::OrphanedNodesManager
+//
+//==============================================================================
+
+PrimUpdaterManager::OrphanedNodesManager::OrphanedNodesManager(
+    const Ufe::Trie<PullVariantInfo>& pulledPrims)
+    : _pulledPrims(pulledPrims)
+{
+}
+
+void PrimUpdaterManager::OrphanedNodesManager::operator()(const Ufe::Notification& n)
+{
+    const auto& sceneNotification = static_cast<const Ufe::SceneChanged&>(n);
+    auto        changedPath = sceneNotification.changedPath();
+
+    // No changedPath means composite.  Use containsDescendant(), as
+    // containsDescendantInclusive() would mean a structure change on the
```

Comment:
Is this micro-optimizatin worth it? What if studio build tooling to change variant in-bulk (which could affect a pulled node) for their pipeline? We're only saving a single comparison of the root of a whole prim sub-tree.